File: simple_automations/os_operations/javascript/directory_clean.js
// npm install argparse
const path = require('path');
const fs = require('fs');
const { ArgumentParser } = require('argparse');

const parser = new ArgumentParser({ description: 'Clean up directory and put files into according folders.' });
parser.addArgument('--path', {
  type: 'string',
  defaultValue: '.',
  help: 'Directory path of the to be cleaned directory',
});

const args = parser.parseArgs();
// rename path variable to dirPath here
const { path: dirPath } = args;

const dir = fs.readdirSync(dirPath);
const files = dir.filter(f => fs.statSync(f).isFile());
const folders = dir.filter(f => fs.statSync(f).isDirectory());

let moved = 0;

console.log(`Cleaning up ${files.length} of ${dir.length} elements.`);

const createdFolders = [];
for (const doc of files) {
  const file = path.parse(doc);

  // skip this script
  if (file.name == 'directory_clean') continue;

  // get name of subfolder to create
  const subfolderPath = path.join(dirPath, file.ext.substring(1));
  
  // create folder if it does not exist
  if (
    !folders.find(f => f == subfolderPath) &&
    !createdFolders.find(f => f == subfolderPath) &&
    !fs.existsSync(subfolderPath)
  ) {
    fs.mkdirSync(subfolderPath);
    createdFolders.push(subfolderPath);
    console.log(`Folder: ${subfolderPath} created.`);
  }

  // move the file to the correct folder
  fs.renameSync(doc, path.join(subfolderPath, file.base));
  moved++;
}

console.log(`Moved ${moved} of ${files.length} files.`);

File: simple_automations/os_operations/python/directory_clean.py
import os
import argparse

parser = argparse.ArgumentParser(
    description="Clean up directory an put files into according folders."
)

parser.add_argument(
    "--path",
    type=str,
    default=".",
    help="Directory path of the to be cleaned up directory"
)

args = parser.parse_args()
path = args.path

print(f"Cleaning up directory {path}")

# get all files from the given directory
dir_content = os.listdir(path)
path_dir_content = [os.path.join(path, doc) for doc in dir_content]
docs = [doc for doc in path_dir_content if os.path.isfile(doc)]
folders = [folder for folder in path_dir_content if os.path.isdir(folder)]
moved = 0
created_folders = []

print(f"Cleaning up {len(docs)} of {len(dir_content)} elements.")

# go through all files and move them into according folders
for doc in docs:
    # separate name from file extension
    full_doc_path, filetype = os.path.splitext(doc)
    doc_path = os.path.dirname(full_doc_path)
    doc_name = os.path.basename(full_doc_path)

    # skip this python file and all hidden files
    if doc_name == "directory_clean" or doc_name.startswith('.'):
        continue

    # get the subfolder name and create the folder if not exist
    subfolder_path = os.path.join(path, filetype[1:].lower())

    if subfolder_path not in folders and subfolder_path not in created_folders:
        try:
            os.mkdir(subfolder_path)
            created_folders.append(subfolder_path)
            print(f"Folder {subfolder_path} created.")
        except FileExistsError as err:
            print(f"Folder already exists at {subfolder_path}... {err}")

    # get the new folder path and move the file
    new_doc_path = os.path.join(subfolder_path, doc_name) + filetype
    os.rename(doc, new_doc_path)
    moved += 1

    print(f"Moved file {doc} to {new_doc_path}")

print(f"Renamed {moved} of {len(docs)} files.")